File: python/ch01/fibonacci/fibonacci/fibonacci_3.py
from typing import Dict

memo: Dict[int, int] = {0: 0, 1: 1}


def fibonacci(n: int) -> int:
  if n not in memo:
    memo[n] = fibonacci(n - 1) + fibonacci(n - 2)

  return memo[n]


def run():
  print("***** Algorithm: HashMap Memoization *****")

  print(f"fibonacci(0) = {fibonacci(0)}")
  print(f"fibonacci(1) = {fibonacci(1)}")
  print(f"fibonacci(5) = {fibonacci(5)}")
  print(f"fibonacci(20) = {fibonacci(20)}")
  print(f"fibonacci(50) = {fibonacci(50)}")

File: rust/ch01/fibonacci/src/fibonacci_3.rs
use std::collections::HashMap;

pub fn fibonacci(n: usize) -> usize {
  let mut memo: HashMap<usize, usize> = HashMap::from([(0, 0), (1, 1)]);
  get_fibonacci(n, &mut memo)
}

fn get_fibonacci(
  n: usize,
  memo: &mut HashMap<usize, usize>,
) -> usize {
  if memo.get(&n).is_none() {
    let value: usize = get_fibonacci(n - 1, memo) + get_fibonacci(n - 2, memo);
    memo.insert(n, value);
  }
  memo.get(&n).unwrap().clone()
}

pub fn run() {
  println!("***** Algorithm: HashMap Memoization *****");

  println!("fibonacci(0) = {}", fibonacci(0));
  println!("fibonacci(1) = {}", fibonacci(1));
  println!("fibonacci(5) = {}", fibonacci(5));
  println!("fibonacci(20) = {}", fibonacci(20));
  println!("fibonacci(50) = {}", fibonacci(50));
}
